File: pluck/routes/info.py
"""POST /api/info — metadata + available qualities (auth-gated, rate-limited)."""
from fastapi import APIRouter, HTTPException, Request

from ..models import InfoReq
from ..mythos import consumer
from ..ratelimit import limiter
from ..ytdlp import extract_info

router = APIRouter()


@router.post("/api/info")
@limiter.limit("30/minute")
def api_info(req: InfoReq, request: Request):
    consumer(request)  # AUTH gate
    url = (req.url or "").strip()
    if not url:
        raise HTTPException(400, "Paste a video URL.")
    try:
        return extract_info(url)
    except ValueError as e:
        raise HTTPException(422, f"Couldn't read that link: {e}")

File: node/src/routes/info.js
/** POST /api/info — metadata + available qualities (auth-gated, rate-limited).
 * Mirrors pluck/routes/info.py. */
import { Router } from "express";
import { rateLimit } from "express-rate-limit";

import { HttpError } from "../mythos.js";
import { extractInfo } from "../ytdlp.js";
import { wrap } from "./helpers.js";

export function createRouter(deps) {
  const { consumer } = deps;
  const router = Router();

  // 30/minute per client address, like the slowapi limiter on the Python route.
  const limiter = rateLimit({
    windowMs: 60_000,
    limit: 30,
    standardHeaders: true,
    legacyHeaders: false,
    message: { detail: "Rate limit exceeded: 30 per 1 minute" },
  });

  router.post("/api/info", limiter, wrap(async (req, res) => {
    consumer(req); // AUTH gate
    const url = String(req.body?.url ?? "").trim();
    if (!url) throw new HttpError(400, "Paste a video URL.");
    try {
      res.json(await extractInfo(url));
    } catch (e) {
      throw new HttpError(422, `Couldn't read that link: ${e.message}`);
    }
  }));

  return router;
}
